Tools/GNUMake/comps/gnu.mak

```diff
@@ -13,23 +13,23 @@ F90FLAGS =
 
 ########################################################################
 
-gcc_version       := $(shell $(CXX) -dumpfullversion -dumpversion | head -1 | sed -e 's;.*  *;;')
-gcc_major_version := $(shell $(CXX) -dumpfullversion -dumpversion | head -1 | sed -e 's;.*  *;;' | sed -e 's;\..*;;')
-gcc_minor_version := $(shell $(CXX) -dumpfullversion -dumpversion | head -1 | sed -e 's;.*  *;;' | sed -e 's;[^.]*\.;;' | sed -e 's;\..*;;')
+gcc_version       = $(shell $(CXX) -dumpfullversion -dumpversion | head -1 | sed -e 's;.*  *;;')
```

Comment:
This change breaks what I'm doing on the GPU branch -- was the immediate expansion of these variables removed for a particular reason?

Reply:
This was intended to fix a flaw in the make system.  If `CXX` is redefined in `Make.local`, we would like it to behave as if it is defined with the overriding value in say `gnu.mak`.  With immediate expansion, unfortunately `gcc_version` etc. still have the old value.   So we changed it.

How does this break the GPU branch?  Could we find a way around this?

Reply:
nvcc has not historically had good support for PGI and XL as host C++ compilers, but we do need them to compile CUDA Fortran. So I opted to go with COMP=PGI or COMP=IBM for compiling GPU code with USE_CUDA=TRUE, but when this is done, we use nvcc + g++ as the C++ compiler. To do this, in pgi.mak, I do an include of gnu.mak so that I get the correct compiler flags for g++ in CXXFLAGS, and then set CXX to nvcc and pass the CXXFLAGS as an option to nvcc. 

The problem this creates without immediate expansion is that DEFINES += -DBL_GCC_VERSION='$(gcc_version)' is evaluated on each compile statement, but CXX no longer is gcc by that point, so it doesn't work, because -dumpversion and -dumpfullversion aren't recognized by non-gcc compilers. (Weirdly though, it seems to be pgc++ that is being invoked with -dumpversion, which should not be CXX at this point. I don't understand that.)

Reply:
I think I can fix this by doing DEFINES := $(DEFINES) after I do the include of gnu.mak, to force immediate expansion. That would break any later updates in Make.local though, I guess?

Reply:
To be clear, it would only break Make.local updates when using CUDA.